proto/api_v2/metricsquery.proto

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2021 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax="proto3";
+
+package jaeger.api_v2;
+
+import "model.proto";
+import "gogoproto/gogo.proto";
+import "google/api/annotations.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/duration.proto";
+import "opentelemetry/proto/metrics/v1/metrics.proto";
+
+option go_package = "api_v2";
+option java_package = "io.jaegertracing.api_v2";
+
+// MetricsQueryBaseRequest is the base request parameter accompanying a MetricsQueryService RPC call.
+message MetricsQueryBaseRequest {
+  // end_time is the ending time of the time series query range.
+  // Optional. Default = now.
+  google.protobuf.Timestamp end_time = 1 [
+    (gogoproto.stdtime) = true,
+    (gogoproto.nullable) = true
+  ];
+
+  // lookback is the duration from the end_time to look back on for metrics data points.
+  // For example, if set to 1h, the query would span from end_time-1h to end_time.
+  // Optional. Default = 1h.
+  google.protobuf.Duration lookback = 2 [
+    (gogoproto.stdduration) = true,
+    (gogoproto.nullable) = true
+  ];
+
+  // step size is the duration between data points of the query results.
+  // For example, if set to 5s, the results would produce a data point every 5 seconds
+  // from the start_time to end_time.
+  // Optional. Default = 5s.
+  google.protobuf.Duration step = 3 [
+    (gogoproto.stdduration) = true,
+    (gogoproto.nullable) = true
+  ];
+
+  // rate is the duration in which the per-second rate of change is calculated for a cumulative counter metric.
+  // Optional. Default = 10m.
+  google.protobuf.Duration rate = 4 [
+    (gogoproto.stdduration) = true,
+    (gogoproto.nullable) = true
+  ];
+}
+
+// GetMetricsResponse is the MetricsQueryService response returned by RPC calls.
+message GetMetricsResponse {
+  // metrics is the collection of metrics returned in a response.
+  repeated opentelemetry.proto.metrics.v1.Metric metrics = 1 [
+    (gogoproto.nullable) = false
+  ];
+}
+
+// GetServiceLatenciesRequest contains parameters for the GetServiceLatencies RPC call.
+message GetServiceLatenciesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+  // quantile is the quantile to compute from latency histogram metrics.
+  // Required.
+  double quantile = 2;
+}
+
+// GetServiceCallRatesRequest contains parameters for the GetServiceCallRates RPC call.
+message GetServiceCallRatesRequest {
```

Comment:
Which service?

Reply:
How do you envision using this API? Eg in an architecture with thousands of services?  It seems you want to pass service names as args, not expect to receive all services.

Reply:
Firstly, I'll take a step back and clarify some sample use cases I have in mind that apply to this API:
- An org-level error rate SLA alert is triggered. How can we root cause this?
  - What services have higher error rates? Are their QPS's significant enough for us to care about?
  - Once we've identified the candidate services, we can drill down into their operations, then further to their traces to pinpoint the child services/operations that are raising the errors and mitigate accordingly.

- Capacity planning:
  - What services have the highest QPSs and latencies? These could be candidate services to target for measuring the bulk of resource consumers.
  - Drilling down into their traces, we can identify the downstream services that incur the bulk of the time and data stores they use.
  - We could view long-term patterns of QPSs and latencies for these downstream services/data stores and anticipate how much more capacity is required for these resource-intensive services.
  - At the same time, we may even identify optimizations on these resource hungry services.

In both cases, we can't easily know, upfront, the service/s with the highest latencies, error or request rates; instead, relying on a sorted view of aggregated metrics grouped by service, which is what I envision this API to provide.